sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,535 @@
+      this.database = database;
+      this.collection = collection;
+      this.filter = filter;
+      this.numberSplit = numberSplit;
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(String.class);
+    }
+
+    @Override
+    public void validate() {
+      Preconditions.checkNotNull(uri, "uri");
+      Preconditions.checkNotNull(database, "database");
+      Preconditions.checkNotNull(collection, "collection");
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) {
+      return false;
+    }
+
+    @Override
+    public BoundedReader createReader(PipelineOptions options) {
+      return new BoundedMongoDbReader(uri, database, collection, filter);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
+      Long estimatedByteSize = 0L;
+      Long totalCfByteSize = 0L;
+
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+      MongoCollection mongoCollection = mongoDatabase.getCollection(collection);
+
+      // get the stats object
+      BasicDBObject stat = new BasicDBObject();
+      stat.append("collStats", collection);
+      Document stats = mongoDatabase.runCommand(stat);
+      totalCfByteSize = Long.valueOf(stats.get("size").toString());
+      avgSize = Long.valueOf(stats.get("avgObjSize").toString());
+      if (filter != null && !filter.isEmpty()) {
+        try {
+          estimatedByteSize = getFilterResultByteSize(mongoCollection, avgSize);
+        } catch (Exception e) {
+          e.printStackTrace();
+        }
+      } else {
+        estimatedByteSize = totalCfByteSize;
+      }
+      return estimatedByteSize;
+    }
+
+    private Long getFilterResultByteSize(MongoCollection mongoCollection, Long avgSize)
+        throws Exception {
+      Long totalFilterByteSize = 0L;
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
+      totalFilterByteSize = countFilter * avgSize;
+      return totalFilterByteSize;
+    }
+
+    @Override
+    public List<BoundedSource> splitIntoBundles(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      Long numberSplitsCalculated = 0L;
+      List<BoundedSource> listEntireCollection = new ArrayList<>();
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+      Long collectionEstimatedSize = getEstimatedSizeBytes(options);
+      ArrayList splitIDFromMongo;
+      double initialRatio = 1.6;
+      if (numberSplit > 0) {
+        numberSplitsCalculated = new Long(numberSplit);
+      } else if (desiredBundleSizeBytes > 0) {
+        numberSplitsCalculated = collectionEstimatedSize / desiredBundleSizeBytes;
+        if (numberSplitsCalculated <= 0) {
+          numberSplitsCalculated = 1L;
```

Comment:
Explain why in this case 1 is the proper result? E.g. you can get collectionEstimatedSize=0 if an error occurs while estimating the size - then is 1 still the proper split granularity to choose?

Either way, please add some comments explaining the reasoning, and ideally some logging too. As a user of this class, I would appreciate knowing how much parallelism the connector chose or was able to achieve.